blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,246 @@
+	TxPool txPoolInterface
+}
+
+type blockchainInterface interface {
+	// WriteBlock writes the block to the db but does not process it
+	WriteBlock(b *types.Block, source string) error
+}
+
+type txPoolInterface interface {
+	Prepare()
+	Peek() *types.Transaction
+	Pop(tx *types.Transaction)
+	Drop(tx *types.Transaction)
+	Demote(tx *types.Transaction)
+}
+
+func NewBlockBuilder(params *BlockBuilderParams) (*BlockBuilder, error) {
+	// vanity extra can only be 32 size max. it is better to trim that to return
+	// an error that we have to propagate. It should be up to higher level
+	// code to error if the vanity extra supplied by the user is too big.
+	if len(params.VanityExtra) > 32 {
+		params.VanityExtra = params.VanityExtra[:32]
+	}
+
+	if params.VanityExtra == nil {
+		params.VanityExtra = make([]byte, 0)
+	}
+
+	if params.Logger == nil {
+		params.Logger = hclog.NewNullLogger()
+	} else {
+		params.Logger = params.Logger.Named("block-builder")
+	}
+
+	builder := &BlockBuilder{
+		params: params,
+	}
+	if err := builder.Reset(); err != nil {
+		return nil, err
+	}
+
+	return builder, nil
+}
+
+type BlockBuilder struct {
+	// input params for the block
+	params *BlockBuilderParams
+
+	// header is the header for the block being created
+	header *types.Header
+
+	// transactions are the data included in the block
+	txns []*types.Transaction
+
+	// block is a reference to the already built block
+	block *types.Block
+
+	state *state.Transition
+}
+
+func (b *BlockBuilder) Reset() error {
+	// Generate the base block
+	header := &types.Header{
+		ParentHash: b.params.Parent.Hash,
+		Number:     b.params.Parent.Number + 1,
+		Miner:      types.ZeroAddress.Bytes(),
+		Nonce:      types.Nonce{},
+		MixHash:    signer.IstanbulDigest,
+		Difficulty: b.params.Parent.Number + 1,
+		StateRoot:  types.EmptyRootHash,
+		Sha3Uncles: types.EmptyUncleHash,
+		GasLimit:   b.params.Parent.GasLimit,
+	}
+
+	b.block = nil
+	b.txns = []*types.Transaction{}
+
+	// calculate gas limit based on parent header
+	header.GasLimit = calculateGasLimit(b.params.Parent.GasLimit, b.params.BlockGasTarget)
+
+	transition, err := b.params.Executor.BeginTxn(b.params.Parent.StateRoot, b.header, b.params.Coinbase)
+	if err != nil {
+		return err
+	}
+
+	b.state = transition
+
+	return nil
+}
+
+// Fill fills the block with transactions from the txpool
+func (b *BlockBuilder) Fill() {
+	b.params.TxPool.Prepare()
+
+	for {
+		tx := b.params.TxPool.Peek()
+
+		// execute transactions one by one
+		finished, err := b.writeTxPoolTransaction(tx)
+		if err != nil {
+			b.params.Logger.Debug("Fill transaction error", "hash", tx.Hash, "err", err)
+		}
+
+		if finished {
+			return
+		}
+	}
+}
+
+func (b *BlockBuilder) writeTx(tx *types.Transaction) error {
+	if err := b.state.Write(tx); err != nil {
+		return err
+	}
+
+	b.txns = append(b.txns, tx)
+
+	return nil
+}
+
+func (b *BlockBuilder) writeTxPoolTransaction(tx *types.Transaction) (bool, error) {
+	if tx == nil {
```

Comment:
Could you move this check to the Fill method and add logging? 
If we have nil transactions there - it shouldnt be silent